authorization/token/service/token_service.go

```diff
@@ -320,6 +325,164 @@ func (s *tokenServiceImpl) Audit(ctx context.Context, identity *account.Identity
 	return &signedToken, nil
 }
 
+// Refresh checks the resource permissions in the given tokenString for the given user, and returns a
+// new RPToken (with a new expiry time and updated permissions if needed)
+func (s *tokenServiceImpl) Refresh(ctx context.Context, identity *account.Identity, accessToken string) (string, error) {
+	log.Debug(ctx, map[string]interface{}{"identity_id": identity.ID.String()}, "refreshing a user token...")
+	// Get the token manager from the context
+	manager, err := token.ReadManagerFromContext(ctx)
+	if err != nil {
+		return "", errors.NewInternalError(ctx, err)
+	}
+	// Now parse the token string that was passed in
+	accessTokenClaims, err := manager.ParseToken(ctx, accessToken)
+	if err != nil {
+		return "", errors.NewBadParameterErrorFromString("access_token", "<hidden>", "failed to parse the access token")
```

Comment:
This looks like BadParameter(..) to this service method, but the controller returning a 400 for an incorrect token looks a bit odd. I'm thinking from a key rotation perspective. A user who hasn't been around for a while comes back and her browser attempts to refresh token. Now since the keys have moved on - the parsing fails and she needs to login again. Getting a 400 might not be very indicative there. @xcoulon @sbryzak What do you think?

Reply:
@sbose78 fixed in https://github.com/fabric8-services/fabric8-auth/commit/f43c9d5